kartothek/io/dask/dataframe_cube.py

```diff
@@ -19,49 +37,94 @@
-    data, cube, store, metadata=None, overwrite=False, partition_on=None
-):
+    data: Union[dd.DataFrame, Dict[str, dd.DataFrame]],
+    cube: Cube,
+    store: Callable[[], simplekv.KeyValueStore],
+    metadata: Optional[Dict[str, Dict[str, Any]]] = None,
+    overwrite: bool = False,
+    partition_on: Optional[Dict[str, Iterable[str]]] = None,
+    shuffle: bool = False,
+    num_buckets: int = 1,
+    bucket_by: Optional[Iterable[str]] = None,
+) -> Delayed:
     """
     Create dask computation graph that builds a cube with the data supplied from a dask dataframe.
 
     Parameters
     ----------
-    data: Union[dask.DataFrame, Dict[str, dask.DataFrame]
+    data:
         Data that should be written to the cube. If only a single dataframe is given, it is assumed to be the seed
         dataset.
-    cube: kartothek.core.cube.cube.Cube
+    cube:
         Cube specification.
-    store: Callable[[], simplekv.KeyValueStore]
+    store:
         Store to which the data should be written to.
-    metadata: Optional[Dict[str, Dict[str, Any]]]
+    metadata:
         Metadata for every dataset.
-    overwrite: bool
+    overwrite:
         If possibly existing datasets should be overwritten.
-    partition_on: Optional[Dict[str, Iterable[str]]]
+    partition_on:
         Optional parition-on attributes for datasets (dictionary mapping :term:`Dataset ID` -> columns).
         See :ref:`Dimensionality and Partitioning Details` for details.
 
     Returns
     -------
-    metadata_dict: dask.Delayed
+    metadata_dict:
         A dask delayed object containing the compute graph to build a cube returning the dict of dataset metadata
         objects.
     """
-    data, ktk_cube_dataset_ids = _ddfs_to_bag(data, cube)
+    check_store_factory(store)
+    if not isinstance(data, dict):
+        data = {cube.seed_dataset: data}
 
-    return (
-        build_cube_from_bag_internal(
-            data=data,
+    ktk_cube_dataset_ids = sorted(data.keys())
+
+    metadata = check_provided_metadata_dict(metadata, ktk_cube_dataset_ids)
+    existing_datasets = discover_datasets_unchecked(cube.uuid_prefix, store)
+    check_datasets_prebuild(ktk_cube_dataset_ids, cube, existing_datasets)
+    partition_on_checked = prepare_ktk_partition_on(
+        cube, ktk_cube_dataset_ids, partition_on
+    )
+    del partition_on
+
+    dct = {}
+    for table_name, ddf in data.items():
+        check_user_df(table_name, ddf, cube, set(), partition_on_checked[table_name])
+
+        indices_to_build = set(cube.index_columns) & set(ddf.columns)
+        if table_name == cube.seed_dataset:
+            indices_to_build |= set(cube.dimension_columns)
+        indices_to_build -= set(partition_on_checked[table_name])
+
+        ddf = ddf.map_partitions(
+            assert_dimesion_index_cols_notnull,
+            ktk_cube_dataset_id=table_name,
             cube=cube,
+            partition_on=partition_on_checked[table_name],
+            meta=ddf._meta,
+        )
```

Comment:
@marco-neumann-by Regarding your other comment about the index null check

Reply:
This was the easiest way to not screw up everything else. Eventually we should consolidate the kartothek dataset behaviour by either providing full nullable support or raising there as well. Deferred this to another PR at a later time

Reply:
> Eventually we should consolidate the kartothek dataset behaviour by either providing full nullable support or raising there as well.

I would rather raise now (aka be more restrictive) and harden + enable later (aka implement a feature) than having a half-tested inconsistent solution now and a potential really ugly breaking change later (aka the other way around). 

Some background: The actual reason why this if forbidden in the cube functionality is not because I deemed it to be an anti-feature but because I wasn't able to build a properly tested feature around it. I also wasn't concerned too much about the index data structure, but about the usage (aka within conditions) where NULL can mean different things, here for non-seed tables:

- A: index data is actual NULL
- B: (very similar) non-indexed payload data is actual NULL
- C: data absent in physical partition list (=primary index)
- D: data absent in secondary index
- E: data absent within parquet file (but both the index and/or the physical partition list hints it could be there), only happens when secondary indices are outdated
- F: data cut out due to table merge (e.g. `(C(column_in_table_a) != 20) & (C(column_in_table_b).isna())`, where both tables have data corresponding to `C(column_in_table_a) == 20`).

Now all these cases are slightly different. Most of them lead to NULL in the resulting dataframe, but the the are created at very different stages of the query processing and getting all these cases right requires some thought and additions to the query system (e.g. the way it is designed, you cannot trivially add case F with the current design).

Reply:
Raising now in kartothek single-tables is a hard breaking change which might surprise users. I'm not willing to deal with this just now. The changes as proposed by this PR would not change any behaviour.

Whether or not it is possible to provide this as a full fledged feature is something I cannot tell at the moment and I don't intend to introduce something like this lightly. Either way, this is out of scope for this change.

Reply:
This is not about single-table. The whole change here concerns cubes and there I would like to know why there is a different behavior between the bag+eager and the DDF interface.

Reply:
clarified offline